src/textual/devtools.py

```diff
@@ -0,0 +1,265 @@
+from __future__ import annotations
+
+import asyncio
+import base64
+import json
+import pickle
+import sys
+import weakref
+from asyncio import Queue, CancelledError
+from contextlib import suppress
+from datetime import datetime, timezone
+from json import JSONDecodeError
+from pathlib import Path
+from typing import cast, Iterable
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+from aiohttp import WSMessage, WSMsgType, WSCloseCode
+from aiohttp.web import run_app
+from aiohttp.web_app import Application
+from aiohttp.web_request import Request
+from aiohttp.web_routedef import get
+from aiohttp.web_ws import WebSocketResponse
+from dateutil.tz import tz
+from rich.align import Align
+from rich.console import Console, ConsoleOptions, RenderResult
+from rich.markup import escape
+from rich.rule import Rule
+from rich.segment import Segments, Segment
+from rich.table import Table
+
+DEFAULT_PORT = 8081
+DEFAULT_SIZE_CHANGE_POLL_DELAY_SECONDS = 2
+QUEUEABLE_TYPES = {"client_log", "client_spillover"}
+
+
+class DevtoolsLogMessage:
+    def __init__(
+        self,
+        segments: Iterable[Segment],
+        path: str,
+        line_number: int,
+        unix_timestamp: int,
+    ) -> None:
+        self.segments = segments
+        self.path = path
+        self.line_number = line_number
+        self.unix_timestamp = unix_timestamp
+
+    def __rich_console__(
+        self, console: Console, options: ConsoleOptions
+    ) -> RenderResult:
+        local_time = (
+            datetime.fromtimestamp(self.unix_timestamp)
+            .replace(tzinfo=timezone.utc)
+            .astimezone(tz=tz.tzlocal())
+        )
+        timezone_name = local_time.tzname()
+        table = Table.grid(expand=True)
+        table.add_column()
+        table.add_column()
+        file_link = f"file://{Path(self.path).absolute()}"
+        file_and_line = f"{Path(self.path).name}:{self.line_number}"
+        table.add_row(
```

Comment:
There is a possibility of console markup injection here. e.g. you could have a path with `[blink]` in it. Can you either use Text objects to build longer strings or call markup.escape on the variables in the f strings.

Reply:
Ah good catch. I caught that elsewhere in the PR but missed it here 😄